docs/design/indexed-job.md

```diff
@@ -0,0 +1,891 @@
+- `--per-completion-env=` implies the jobs `.spec.completions` is set to the length of the argument array (3 in the example).
+- `--per-completion-env=F=<values>` causes env var with `F` to be available in the enviroment when the command line is evaluated.
+
+How exactly this happens is discussed later in the doc: this is a sketch of the user experience.
+
+In practice, the list of files might be much longer and stored in a file
+on the users local host, like:
+
+```
+$ cat files-to-process.txt
+12342.dat
+97283.dat
+38732.dat
+...
+```
+
+So, the user could specify instead: `--per-completion-env=F="$(cat files-to-process.txt)"`.
+
+However, `kubectl` should also support a format like:
+ `--per-completion-env=F=@files-to-process.txt`.
+That allows `kubectl` to parse the file, point out any syntax errors, and would not run up against command line length limits (2MB is common, as low as 4kB is POSIX compliant).
+
+One case we do not try to handle is where the file of work is stored on a cloud filesystem, and not accessible from the users local host.  Then we cannot easily use indexed job, because we do not know the number of completions.  In this case, the user could use the Work-Queue style of Job (already supported).
+
+Another case we do not try to handle is where the input file does not exist yet because this Job is to be run at a future time, or depends on another job.   The workflow and scheduled job proposal need to consider this case.
+
+#### Multiple parameters
+
+The user may also have multiple paramters, like in [work list 2](#work-list-2).
+One way is to just list all the command lines already expanded, one per line, in a file, like this:
+
+```
+$ cat matrix-commandlines.txt
+/usr/local/bin/process_matrix_block -start_row 0 -end_row 15 -start_col 0 --end_col 15
+/usr/local/bin/process_matrix_block -start_row 16 -end_row 31 -start_col 0 --end_col 15
+/usr/local/bin/process_matrix_block -start_row 0 -end_row 15 -start_col 16 --end_col 31
+/usr/local/bin/process_matrix_block -start_row 16 -end_row 31 -start_col 16 --end_col 31
+```
+
+and run the Job like this:
+
+```
+kubectl run process-matrix --image=my/matrix \
+   --per-completion-env=COMMAND_LINE=@matrix-commandlines.txt \
+   --restart=OnFailure  \
+   -- \
+   'eval "$COMMAND_LINE"'
+```
+
+However, this may have some subtleties with shell escaping.  Also, it depends on the user
+knowing all the correct arguments to the docker image being used (more on this later).
+
+Instead, kubectl should support multiple instances of the `--per-completion-env` flag.  For example, to implement work list 2, a user could do:
+
+```
+kubectl run process-matrix --image=my/matrix \
+   --per-completion-env=SR="0 16 0 16" \
+   --per-completion-env=ER="15 31 15 31" \
+   --per-completion-env=SC="0 0 16 16" \
+   --per-completion-env=EC="15 15 31 31" \
+   --restart=OnFailure  \
+   -- \
+   /usr/local/bin/process_matrix_block -start_row $SR -end_row $ER -start_col $ER --end_col $EC 
```

Comment:
Hmm.  That raises some sticky issues.
The `$(VAR)` style is for kubelet-substitution of env vars.  If we chose the server side template expansion (multiple parameters), then you would use `$(VAR)`.  But if we chose the index-only approach, then it needs to be shell-substitution, which is just `$VAR`.  For shell-substitution, dollar-paren means run a command, which is the wrong thing entirely.